substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1011 @@
+				next_auth.len() as u32,
+			);
+
+			// Get next epoch params
+			let randomness = NextRandomness::<T>::get();
+			let epoch_idx = EpochIndex::<T>::get() + 1;
+
+			let mut segment = BoundedVec::with_max_capacity();
+			for ticket in tickets {
+				debug!(target: LOG_TARGET, "Checking ring proof");
+
+				let ticket_id_input =
+					vrf::ticket_id_input(&randomness, ticket.body.attempt_idx, epoch_idx);
+				let Some(ticket_id_output) = ticket.signature.outputs.get(0) else {
+					debug!(target: LOG_TARGET, "Missing ticket vrf output from ring signature");
+					continue
+				};
+				let ticket_id = vrf::make_ticket_id(&ticket_id_input, &ticket_id_output);
+				if ticket_id >= ticket_threshold {
+					debug!(target: LOG_TARGET, "Over threshold");
+					continue
+				}
+
+				let sign_data = vrf::ticket_body_sign_data(&ticket.body, ticket_id_input);
+
+				if ticket.signature.verify(&sign_data, &verifier) {
+					TicketsData::<T>::set(ticket_id, Some(ticket.body));
+					segment
+						.try_push(ticket_id)
+						.expect("has same length as bounded input vector; qed");
+				} else {
+					debug!(target: LOG_TARGET, "Proof verification failure");
+				}
+			}
+
+			if !segment.is_empty() {
+				debug!(target: LOG_TARGET, "Appending segment with {} tickets", segment.len());
+				segment.iter().for_each(|t| debug!(target: LOG_TARGET, "  + {t:16x}"));
+				let mut metadata = TicketsMeta::<T>::get();
+				NextTicketsSegments::<T>::insert(metadata.segments_count, segment);
+				metadata.segments_count += 1;
+				TicketsMeta::<T>::set(metadata);
+			}
+
+			Ok(())
+		}
+
+		/// Plan an epoch config change.
+		///
+		/// The epoch config change is recorded and will be announced at the begin of the
+		/// next epoch together with next epoch authorities information.
+		/// In other words the configuration will be activated one epoch after.
+		/// Multiple calls to this method will replace any existing planned config change that had
+		/// not been enacted yet.
+		///
+		/// TODO-SASS-P4: proper weight
+		#[pallet::call_index(1)]
+		#[pallet::weight({0})]
+		pub fn plan_config_change(
+			origin: OriginFor<T>,
+			config: EpochConfiguration,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			ensure!(
+				config.redundancy_factor != 0 && config.attempts_number != 0,
+				Error::<T>::InvalidConfiguration
+			);
+			PendingEpochConfigChange::<T>::put(config);
+			Ok(())
+		}
+
+		/// Report authority equivocation.
+		///
+		/// This method will verify the equivocation proof and validate the given key ownership
+		/// proof against the extracted offender. If both are valid, the offence will be reported.
+		///
+		/// This extrinsic must be called unsigned and it is expected that only block authors will
+		/// call it (validated in `ValidateUnsigned`), as such if the block author is defined it
+		/// will be defined as the equivocation reporter.
+		///
+		/// TODO-SASS-P4: proper weight
+		#[pallet::call_index(2)]
+		#[pallet::weight({0})]
```

Comment:
Remove! 
This is not part of the core protocol and should be added in a separate PR